### PR TITLE
Proof-of-concept: Open status links in-app if possible

### PR DESCRIPTION
**Pull request by oneiros**
This is a sketch for an alternative approach to #26917 

The goal is to mimick the behavior of native apps in our web UI and not open links in a new tab/window if we can display them "in-app". This is the case for accounts and statuses. The native apps use the search API to "resolve" the linked URL and then decide how to open it. If we try the same on the web, the search API request might take too long and we trigger the browser's built-in pop-up blocking.

#26917 works around these issues by linking to a local "redirect" controller that resolves the URL and redirects properly. This works well but has two important downsides: "Internal" links will still open in a new/tab window and a new "open redirect" is introduced that could easily be exploited.

We pondered a few possible alternatives, including annotating the HTML server-side. But that would add considerable processing time to an already critical code path.

This is a proof of concept of a different approach: If timing is the problem, then maybe we should try to stay within the bounds *most* of the time. This will not work all of the time, but might be "good enough". Besides fixing the issues mentioned above this could also benefit other clients.

The idea is twofold: To get faster results, I added a new API endpoint to resolve URLs. This works similar to the search API but will return faster if a status / account is already known in the local database. A couple of additional optimizations are possible (see below).

On the client-side I added a timeout to the API call to guarantee we stay within the browser's limits. If the timeout triggers, I open a modal warning the user that they are now leaving the site. If they confirm, the link opens in a new tab/window.

This is still only a rough sketch: Here is an incomplete list of what is still missing:

* This is only for statuses. The same should be implemented for account bios.
* When resolving URLs only make an HTTP call to the external server if its domain is within the list of known instances. This should cut down response times for external links a lot.
* A loading indicator when the link is clicked. Even with a small timeout, users might be worried if a click has no immediate effect.
* Proper naming. I am especially unhappy with the name of the new API endpoint. "Resolved URL" is consisent with the existing naming, but does not express what is really going on very well.
* I currently use the existing confirmation modal. This should maybe be replaced to be consistent with the new "interstitial" introduced in #27792. I could not easily re-use that because it currently lives outside of the react app and cannot handle arbitrary external URLs yet.
* Tests

Open questions:
* Should we increase the timeout? I tested with one second, which is still okay(ish) in terms of UI snappiness, but at least double that should be possible.
* Should we cache the resolved URL in the rails cache? This would make the new API endpoint even faster at the cost of redis memory.
* Does the HTTP caching directive make sense here? (I copy&pasted it from #26917)
* And most importantly: Do we  want to go down this route?

(JS code heavily inspired by #28762. Thanks @Joshix-1!)

### CODE REVIEW

File: app/controllers/api/v2/resolved_urls_controller.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
+class Api::V2::ResolvedUrlsController < Api::BaseController
```
Comment (renchap) [inline app/controllers/api/v2/resolved_urls_controller.rb:3]: Any reason to make this v2? There is no v1 for this API?

File: app/javascript/mastodon/components/status_content.jsx [hunk 2]
```diff
@@ -64,6 +70,20 @@ class TranslateButton extends PureComponent {
...
 
+const mapDispatchToProps = (dispatch, { intl }) => ({
+  openExternalLink(url) {
+    dispatch(openModal({
+      modalType: 'CONFIRM',
+      modalProps: {
+        message: intl.formatMessage(messages.openExternalLink),
+        confirm: intl.formatMessage(messages.openExternalLinkConfirm),
+        closeWhenConfirm: true,
+        onConfirm: () => window.open(url, null, 'norefferer'),
```
Comment (renchap) [inline app/javascript/mastodon/components/status_content.jsx:81]: Should be `noreferrer`

File: app/controllers/api/v2/resolved_urls_controller.rb (hunk 1 above)
Comment (oneiros) [reply app/controllers/api/v2/resolved_urls_controller.rb:3]: No reason, just that I had not fully internalized our policy here.

File: app/javascript/mastodon/components/status_content.jsx (hunk 2 above)
Comment (oneiros) [reply app/javascript/mastodon/components/status_content.jsx:81]: Oops. You are of course absolutely right.

**Comment by ClearlyClaire** [comment]
I think I agree with the idea of performing the check in-app, and falling back to an in-app modal, rather than unconditionally opening a tab that could take a long time to load.

I do have a small concern with this approach, though, in that I'm not sure there is an universal “safe” timeout that we are sure won't trigger the browser's pop-up blocker. If there isn't, I'm not sure how we can offer graceful feedback. But maybe there is such a safe timeout, and this is not an issue.

> The idea is twofold: To get faster results, I added a new API endpoint to resolve URLs. This works similar to the search API but will return faster if a status / account is already known in the local database.

That idea sounds good. A few remarks:
- since we are going for a new endpoint, I think it should be `POST` instead of `GET`, because of two reasons:
  - we've used `POST` a few times for endpoints considered as expensive (which is potentially the case here)
  - this would avoid the link from appearing in the query string and thus the server's logs
- the code you moved to `find_remote_status_in_local_db` is very useful for Mastodon posts, but it may not be reliable at all for other servers with other URL schemes (Friendica, Pleroma, even Threads)… so those are still going to go through the slow path; probably not much we can do about that though

> * Should we increase the timeout? I tested with one second, which is still okay(ish) in terms of UI snappiness, but at least double that should be possible.

The hardest limit will be what the browser allow us, but UX-wise, we can have multiple tiers, e.g. having a “loading” modal pop up after 1 second, and change to a “failed to resolve, click here to leave Mastodon” modal after the longer timeout.

> * Should we cache the resolved URL in the rails cache? This would make the new API endpoint even faster at the cost of redis memory.

I think this could make sense, but we'd have to be careful to still run the permission check. But I'm also a bit worried about this allowing attackers to just fill the cache with arbitrary links.

> * Does the HTTP caching directive make sense here? (I copy&pasted it from

I *think* so, but also, does that feature make sense for people who are logged out?

**Pull request closed without merging**